sw/host/ot_certs/src/asn1/x509.rs

```diff
@@ -0,0 +1,349 @@
+        // rdnSequence  RDNSequence }
+        //
+        // RDNSequence ::= SEQUENCE OF RelativeDistinguishedName
+        //
+        // RelativeDistinguishedName ::=
+        //     SET SIZE (1..MAX) OF AttributeTypeAndValue
+        //
+        // AttributeTypeAndValue ::= SEQUENCE {
+        //     type     AttributeType,
+        //     value    AttributeValue }
+        //
+        // AttributeType ::= OBJECT IDENTIFIER
+        //
+        // AttributeValue ::= ANY -- DEFINED BY AttributeType
+        builder.push_seq(name_hint.clone(), |builder| {
+            for (attr_type, val) in name {
+                builder.push_set(concat_suffix(&name_hint, "set"), |builder| {
+                    builder.push_seq(
+                        concat_suffix(&name_hint, &attr_type.to_string()),
+                        |builder| {
+                            builder.push_oid(&attr_type.oid())?;
+                            builder.push_string(
+                                concat_suffix(&name_hint, &attr_type.to_string()),
+                                &attr_type.data_type(),
+                                val,
+                            )
+                        },
+                    )
+                })?;
+            }
+            Ok(())
+        })
+    }
+
+    pub fn push_public_key_info<B: Builder>(
+        builder: &mut B,
+        pubkey_info: &SubjectPublicKeyInfo,
+    ) -> Result<()> {
+        // From https://datatracker.ietf.org/doc/html/rfc5280#section-4.1:
+        // SubjectPublicKeyInfo  ::=  SEQUENCE  {
+        // algorithm            AlgorithmIdentifier,
+        // subjectPublicKey     BIT STRING  }
+        builder.push_seq(Some("subject_public_key_info".into()), |builder| {
+            Self::push_pubkey_alg_id(builder, pubkey_info)?;
+            builder.push_as_bit_string(
+                Some("subject_public_key".into()),
+                &Tag::BitString,
+                0,
+                |builder| Self::push_public_key(builder, pubkey_info),
+            )
+        })
+    }
+
+    pub fn push_pubkey_alg_id<B: Builder>(
+        builder: &mut B,
+        pubkey_info: &SubjectPublicKeyInfo,
+    ) -> Result<()> {
+        // From https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.1.2
+        // AlgorithmIdentifier  ::=  SEQUENCE  {
+        // algorithm               OBJECT IDENTIFIER,
+        // parameters              ANY DEFINED BY algorithm OPTIONAL  }
+        builder.push_seq(
+            Some("subject_public_key_alg".into()),
+            |builder| match pubkey_info {
+                SubjectPublicKeyInfo::EcPublicKey(ec_pubkey) => {
+                    builder.push_oid(&Oid::EcPublicKey)?;
+                    Self::push_ec_public_key_params(builder, ec_pubkey)
+                }
+            },
+        )
+    }
+
+    pub fn push_public_key<B: Builder>(
+        builder: &mut B,
+        pubkey_info: &SubjectPublicKeyInfo,
+    ) -> Result<()> {
```

Comment:
For all these `push` functions, it might worth having a trait and then instead of
```rust
Self::push_public_key(builder, pubkey_info)
```
you would just do
```rust
builder.push(pubkey_info)
```

But I think we can tweak that later and get this PR merged ASAP.

Reply:
Yes I think a trait would be nicer but there are some subtleties involved with how to handle IMPLICIT tags so it should not just be a straight `push`. Since I didn't have a good understanding of what is needed when I started, I decided not to go for a trait yet. Now that we now what we need to support, we are in a better position to design a trait I think.